src/transactions/TransactionUtils.cpp

```diff
@@ -1142,34 +1142,245 @@ claimableBalanceFlagIsValid(ClaimableBalanceEntry const& cb)
+            auto const& oe = offer.current().data.offer();
+            if (!(oe.sellerID == accountID))
+            {
+                throw std::runtime_error("Offer not owned by expected account");
+            }
+            else if (!(oe.buying == asset || oe.selling == asset))
+            {
+                throw std::runtime_error(
+                    "Offer not buying or selling expected asset");
+            }
+
+            releaseLiabilities(ltxInner, header, offer);
+            auto trustAcc = stellar::loadAccount(ltxInner, accountID);
+            removeEntryWithPossibleSponsorship(ltxInner, header,
+                                               offer.current(), trustAcc);
+            offer.erase();
         }
-        else if (!(oe.buying == asset || oe.selling == asset))
+        ltxInner.commit();
+    }
+
+    // redeem from pools
+    if (ledgerVersion >= 18)
+    {
+        LedgerTxn ltxInner(ltx);
+
+        // load all relevant pool share trustLines
+        auto poolShareTrustLines =
+            ltxInner.loadPoolShareTrustLinesByAccountAndAsset(accountID, asset);
+
+        if (poolShareTrustLines.empty())
+        {
+            return true;
+        }
+
+        // load all relevant pools
+        UnorderedSet<LedgerKey> poolKeys;
+        for (auto& trustLine : poolShareTrustLines)
         {
-            throw std::runtime_error(
-                "Offer not buying or selling expected asset");
+            poolKeys.emplace(liquidityPoolKey(
+                trustLine.current().data.trustLine().asset.liquidityPoolID()));
         }
+        ltxInner.prefetch(poolKeys);
 
-        releaseLiabilities(ltxInner, header, offer);
-        auto trustAcc = stellar::loadAccount(ltxInner, account);
-        removeEntryWithPossibleSponsorship(ltxInner, header, offer.current(),
-                                           trustAcc);
-        offer.erase();
+        for (auto& poolShareTrustLine : poolShareTrustLines)
+        {
+            // use a lambda so we don't hold a reference to the internals of
+            // TrustLineEntry
+            auto poolTL = [&]() -> TrustLineEntry const& {
+                return poolShareTrustLine.current().data.trustLine();
+            };
+
+            auto pool =
+                loadLiquidityPool(ltxInner, poolTL().asset.liquidityPoolID());
+            auto balance = poolTL().balance;
+            auto cbSponsoringAccID =
+                poolTL().ext.v() == 1 &&
+                        getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                            .sponsoringID
+                    ? *getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                           .sponsoringID
+                    : accountID;
+
+            // release reserves for pool share trustline that will be deleted
+            // later
+            {
+                auto trustAcc = stellar::loadAccount(ltxInner, accountID);
+                removeEntryWithPossibleSponsorship(
```

Comment:
It would be really great if we could erase `poolShareTrustLine` at the same time as we release the reserves. As part of #3169, I had to deal with some cases like `ManageOfferOpFrameBase` where we erased at a different place and it is really confusing. Looking at the rest of this function, it seems like this should be feasible.

Reply:
Done